pkg/controller/volume/pvprotection/pv_protection_controller.go

```diff
@@ -0,0 +1,258 @@
+// processNextWorkItem deals with one pvcKey off the queue.  It returns false when it's time to quit.
+func (c *PVProtectionController) processNextWorkItem() bool {
+	pvKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(pvKey)
+
+	_, pvName, err := cache.SplitMetaNamespaceKey(pvKey.(string))
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Error parsing PV key %q: %v", pvKey, err))
+		return true
+	}
+
+	err = c.processPV(pvName)
+	if err == nil {
+		c.queue.Forget(pvKey)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("PV %v failed with : %v", pvKey, err))
+	c.queue.AddRateLimited(pvKey)
+
+	return true
+}
+
+func (c *PVProtectionController) processPV(pvName string) error {
+	glog.V(4).Infof("Processing PV %s", pvName)
+	startTime := time.Now()
+	defer func() {
+		glog.V(4).Infof("Finished processing PV %s (%v)", pvName, time.Now().Sub(startTime))
+	}()
+
+	pv, err := c.pvLister.Get(pvName)
+	if apierrs.IsNotFound(err) {
+		glog.V(4).Infof("PV %s not found, ignoring", pvName)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if isDeletionCandidate(pv) {
+		// PV should be deleted. Check if it's used and remove finalizer if
+		// it's not.
+		isUsed := c.isBeingUsed(pv)
+		if !isUsed {
+			return c.removeFinalizer(pv)
+		}
+	}
+
+	if needToAddFinalizer(pv) {
+		// PV is not being deleted -> it should have the finalizer. The
+		// finalizer should be added by admission plugin, this is just to add
+		// the finalizer to old PVs that were created before the admission
+		// plugin was enabled.
+		return c.addFinalizer(pv)
+	}
+	return nil
+}
+
+func (c *PVProtectionController) addFinalizer(pv *v1.PersistentVolume) error {
+	pvClone := pv.DeepCopy()
+	pvClone.ObjectMeta.Finalizers = append(pvClone.ObjectMeta.Finalizers, volumeutil.PVProtectionFinalizer)
+	_, err := c.client.CoreV1().PersistentVolumes().Update(pvClone)
+	if err != nil {
+		glog.V(3).Infof("Error adding protection finalizer to PV %s: %v", pv.Name)
+		return err
+	}
+	glog.V(3).Infof("Added protection finalizer to PV %s", pv.Name)
+	return nil
+}
+
+func (c *PVProtectionController) removeFinalizer(pv *v1.PersistentVolume) error {
+	pvClone := pv.DeepCopy()
+	pvClone.ObjectMeta.Finalizers = slice.RemoveString(pvClone.ObjectMeta.Finalizers, volumeutil.PVProtectionFinalizer, nil)
+	_, err := c.client.CoreV1().PersistentVolumes().Update(pvClone)
+	if err != nil {
+		glog.V(3).Infof("Error removing protection finalizer from PV %s: %v", pv.Name, err)
+		return err
+	}
+	glog.V(3).Infof("Removed protection finalizer from PV %s", pv.Name)
+	return nil
+}
+
+func (c *PVProtectionController) isBeingUsed(pv *v1.PersistentVolume) bool {
+	// check if PV is being bound to a PVC by its status
+	// the status will be updated by PV controller
+	if pv.Status.Phase == v1.VolumeFailed || pv.Status.Phase == v1.VolumeReleased {
+		// the PV is not being used now
+		return false
+	}
+	if pv.Status.Phase == v1.VolumeAvailable && pv.Spec.ClaimRef == nil {
```

Comment:
Admin can post a pre-bound PV (`pv.Spec.ClaimRef` is not nil) and they should be allowed to delete it until user does not create appropriate PVC. IMO, we could trust `pv.Status.Phase==VolumeAvailable` here without races - PV controller won't bind anything that's marked for deletion.

Please allow deletion of `Pending` volumes too.

Reply:
Now that I think about it, can we leave all the dirty work to PV controller and ignore PVCs completely in this PR? PV controller makes sure that PVs are Released or Failed when they're ready to be deleted...

This protection controller would only ensure that PV that's Bound can't be deleted.

Reply:
Another situation:  a PV is bound to a PVC, but the PVC has deletionTimeStamp set, should we allow admin to delete the PV ?  Or wait util PVC is deleted and PV is not bound ?

Reply:
> a PV is bound to a PVC, but the PVC has deletionTimeStamp set, should we allow admin to delete the PV

No, the PVC is needed by something that holds the finalizer, e.g. a running pod. We must not delete the PV while it's used by a pod!